### PR TITLE
Added Activity type and relevant providers.

### PR DESCRIPTION
**Pull request by BriceCroix**
Added datatypes :
- `Activity`
- `TrackSegment`
- `TrackPoint`
with corresponding models.

Added repository for the Activities, that get stored to an isar database.

Classes diagram is now as follow :
```mermaid
---
title: Movna classes uml diagram
---
classDiagram
    class GpsCoordinates{
        +latitude : Angle
        +longitude : Angle
    }

    class Location{
        +error : Distance
        +altitude : Distance
        +heading : Angle
        +speed : Speed
        +speedError : Speed
    }
    Location *-- "1" GpsCoordinates : +gpsCoordinates

    class TimedLocation{
        +timestamp : DateTime
    }
    TimedLocation *-- "1" Location : +location
    note for TimedLocation "Mirrors geolocator's Position"

    class TrackPoint{
        +timestamp : DateTime
        +heartRate : Frequency?
    }
    TrackPoint *-- "0..1" Location : +location

    class TrackSegment
    TrackSegment *-- "1..*" TrackPoint : +points

    class Activity{
        +startTime : DateTime
        +stopTime : DateTime?
        +sport : Sport?
        +name : String?
        +durationRunning : Duration
        +distance : Distance?
        +averageSpeed : Speed?
        +maxSpeed : Speed?
        +averageHeartRate : Frequency?
        +notes : String?
    }
    Activity *-- "0..*" TrackSegment : +segments
```

### CODE REVIEW

**Comment by BriceCroix** [comment]
There may be some things to rewrite, I am quite unsure of the relevancy of a activity source there, as the repository is quite empty, taking only care of the entity-model conversion. I am open to suggestions ! :)

File: lib/core/unit_conversion.dart [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import 'dart:math';
+
+const _kiloFactor = 1e3;
```
Comment (ronw98) [inline lib/core/unit_conversion.dart:3]: use the packages
https://pub.dev/packages/units_converter
https://pub.dev/packages/fling_units

File: lib/data/adapters/activity_isar_adapter.dart [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+  ActivityIsarAdapter(this.trackSegmentIsarAdapter);
+
+  TrackSegmentIsarAdapter trackSegmentIsarAdapter;
+
+  @override
+  Activity modelToEntity(ActivityModel m) {
+    return Activity(
+      name: m.name,
+      sport: m.sport,
+      // TODO : convert startTime and stopTime back to their original timezone
```
Comment (ronw98) [inline lib/data/adapters/activity_isar_adapter.dart:21]: do not use timezones, always use UTC and then the presentation will account for timezones when displaying information to the user, all dates operations should be made using UTC dates

**Review by ronw98 — CHANGES_REQUESTED** [review]

File: lib/core/injection.dart [hunk 3]
```diff
@@ -1,8 +1,56 @@
...
+    isar = await Isar.open(
+      [ActivityModelSchema],
+      directory: directory.path,
+    );
+  } catch (e) {
+    return const Left(Failure.databaseNotOpened());
+  }
+
+  return Right(isar);
+}
```
[the remark this answers is not in the document]
Comment (BriceCroix) [reply lib/core/injection.dart:56]: I agree on the try catch in the repository, but I don't see another way of using Isar than creating a singleton there. This allows to have a single instance than manages all schemas, and schemas could potentially be linked in the future, it is then important to have have opened at once there.

File: lib/data/adapters/activity_isar_adapter.dart (hunk 2 above)
Comment (BriceCroix) [reply lib/data/adapters/activity_isar_adapter.dart:21]: Yes, UTC should be used to store the dates in the database (even more since I use the startTime as unique ID), but the Datetimes get stored as UTC, and are loaded in localtime by Isar automatically. The problem with that is that if one does an activity at 12h in april, it will be marked as done at 11h (or 13h I don't remember) after the DST change in may or june. The phenomenon is even worse if the user wants to see its activities when they're in another country, it will be marked as done at whatever time it was in their local time when they did the activity in their country of origin.

I think this behavior is ok for now, but it is definitely an issue to address in the future.

File: lib/core/injection.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (ronw98) [reply lib/core/injection.dart]: I added some code to change the Isar initialization using `injectable`, if you want to take a look at it.
The only thing is it forces the app to call `getApplicationDirectory` on startup wich can throw but we can account for it at startup, catching any initialization error and displaying a screen accordingly

File: lib/data/adapters/activity_isar_adapter.dart (hunk 2 above)
Comment (ronw98) [reply lib/data/adapters/activity_isar_adapter.dart:21]: ok I get what you mean, in this particular case we don't want to display the time using current timezone but actually using the activity's timezone so that the user knows at what time of day they actually exercised.

What we could do is simply store the activity's timezone in isar so that the presentation can make the correct adjustments (we could imagine a setting to either display at current timezone or activity timezone)

File: lib/core/unit_conversion.dart [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+/// Converts [angle] in degrees to radians.
+double degreesToRadians(double angle) {
+  return angle * _degreesToRadiansRatio;
+}
+
+/// Converts [angle] in radians to degrees.
+double radiansToDegrees(double angle) {
+  return angle / _degreesToRadiansRatio;
+}
```
Comment (ronw98) [inline lib/core/unit_conversion.dart:38]: I'd rather put all this in a class, either with static members or injected but I don't really like top level methods, this is purely personal I don't think there are general guidelines about this

File: lib/data/adapters/activity_isar_adapter.dart (hunk 2 above)
Comment (ronw98) [reply lib/data/adapters/activity_isar_adapter.dart:21]: also, if you create a TODO, create an issue and link it in the TODO so that we can keep track of that

File: lib/data/adapters/activity_isar_adapter.dart [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+      ..startTime = e.startTime
+      ..stopTime = e.stopTime
+      ..distanceInMeters = e.distanceInMeters
+      ..durationInMicroSeconds = e.duration.inMicroseconds
+      ..averageSpeedInMetersPerSecond = e.averageSpeedInMetersPerSecond
+      ..maxSpeedInMetersPerSecond = e.maxSpeedInMetersPerSecond
+      ..averageHeartBeatPerMinute = e.averageHeartBeatPerMinute
+      ..notes = e.notes
+      ..trackSegments =
+          trackSegmentIsarAdapter.entitiesToModels(e.trackSegments);
```
Comment (ronw98) [inline lib/data/adapters/activity_isar_adapter.dart:48]:
```suggestion
          trackSegmentIsarAdapter.entitiesToModels(e.trackSegments,);
```
+ format

File: lib/data/adapters/activity_isar_adapter.dart [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+      speedErrorInMetersPerSecond: m.speedErrorInMetersPerSecond,
+      timestamp: m.timestamp,
+    );
+  }
+
+  @override
+  LocationModel entityToModel(Location e) {
+    final m = LocationModel()
+      ..gpsCoordinates =
+          gpsCoordinatesIsarAdapter.entityToModel(e.gpsCoordinates)
```
Comment (ronw98) [inline lib/data/adapters/activity_isar_adapter.dart:125]:
```suggestion
          gpsCoordinatesIsarAdapter.entityToModel(e.gpsCoordinates,)
```
+ format

File: lib/data/datasources/isar_database_source.dart [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+  Future<Either<Failure, void>> saveActivity(ActivityModel model) async {
+    final either = await _getIsar();
+    return either.fold((failure) {
+      return Left(failure);
+    }, (isar) async {
+      final id = await isar.writeTxn(() async {
+        return await isar.activityModels.put(model);
+      });
+      if (id != model.id) {
+        return const Left(Failure.databaseNotSaved());
```
Comment (ronw98) [inline lib/data/datasources/isar_database_source.dart:45]: here rather throw than return a `Failure` (when you'll have removed all the `Either`s from the datasource`

File: lib/data/datasources/isar_database_source.dart [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    final either = await _getIsar();
+
+    return either.fold((failure) {
+      return Left(failure);
+    }, (isar) async {
+      final deleted = await isar.writeTxn(() async {
+        return await isar.activityModels.delete(id);
+      });
+      if (!deleted) {
+        return const Left(Failure.databaseNotDeleted());
```
Comment (ronw98) [inline lib/data/datasources/isar_database_source.dart:63]: same here, throw rather than return a failure

File: lib/data/models/activity_model.dart [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+@Collection()
+class ActivityModel {
+  /// Use startTime as unique identifier.
+  Id get id => startTime.microsecondsSinceEpoch;
+
+  DateTime startTime = DateTime.fromMillisecondsSinceEpoch(0);
+
+  DateTime? stopTime;
+
+  // TODO : add field timezoneOffset or timezone to store localTime as well,
```
Comment (ronw98) [inline lib/data/models/activity_model.dart:15]: yup excellent, same as in the adapter though, create an issue and link it (both TODOs will link the same issue)

EDIT: actually might as well add that field already

File: lib/data/repositories/activity_repository_impl.dart [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    try {
+      final either = await _source.getActivities();
+      return either.fold((failure) {
+        return Left(failure);
+      }, (activities) {
+        return Right(_activityAdapter.modelsToEntities(activities));
+      });
+    } catch (e) {
+      return const Left(Failure.database());
+    }
```
Comment (ronw98) [inline lib/data/repositories/activity_repository_impl.dart:52]: rework this when after changing datasource methods signatures

File: lib/data/models/activity_model.dart [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+@embedded
+class TrackSegmentModel {
+  List<TrackPointModel> trackPoints = [];
+}
+
+@embedded
+class TrackPointModel {
+  LocationModel? location;
```
Comment (ronw98) [inline lib/data/models/activity_model.dart:46]: why would the location be null ?

File: lib/domain/entities/activity.dart [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+    /// The sport practiced during this activity. Can be unspecified.
+    Sport? sport,
+
+    /// The name of this activity. Can be unspecified.
+    String? name,
+
+    /// The total distance traveled during this activity in meters.
+    @Default(0) double distanceInMeters,
+
+    /// The total duration of this activity, accounting for pauses.
```
Comment (ronw98) [inline lib/domain/entities/activity.dart:30]: I'm confused by this, if I start the activity at 15:00, end it at 16:00 and make a 15min pause in the middle, will the duration be 45min et 1h?
I think it'd be clearer using `includingPauses` if the result should be 1h or `excludingPauses` otherwise.

The question remaining is, why isn't the other duration available here ?

File: lib/domain/entities/activity.dart [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+    @Default(Duration.zero) Duration duration,
+
+    /// The maximum speed reached during this activity in meters per second.
+    @Default(0) double maxSpeedInMetersPerSecond,
+
+    /// The average speed during this activity in meters per second.
+    @Default(0) double averageSpeedInMetersPerSecond,
+
+    /// The average heart beat during this activity in Beats Per Minute.
+    /// Can be unspecified.
```
Comment (ronw98) [inline lib/domain/entities/activity.dart:40]: prefer using `null` rather than `unspecified`

for example the field `averageSpeedInMetersPerSecond` can be unspecified but it cannot be null

File: lib/domain/entities/activity.dart [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+
+  const factory Activity({
+    /// The time at which this activity begun.
+    required DateTime startTime,
+
+    /// The time at which this activity ended.
+    /// Null when activity is not ended yet.
+    DateTime? stopTime,
+
+    /// The sport practiced during this activity. Can be unspecified.
```
Comment (ronw98) [inline lib/domain/entities/activity.dart:21]: prefer using null rather than unspecified

for example the field averageSpeedInMetersPerSecond can be unspecified but it cannot be null

File: lib/domain/entities/activity.dart [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+@freezed
+class Activity with _$Activity {
+  const Activity._();
+
+  const factory Activity({
+    /// The time at which this activity begun.
+    required DateTime startTime,
+
+    /// The time at which this activity ended.
+    /// Null when activity is not ended yet.
```
Comment (ronw98) [inline lib/domain/entities/activity.dart:18]:
```suggestion
    /// Null while activity has not ended.
```

File: lib/domain/entities/activity.dart [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+    /// The average speed during this activity in meters per second.
+    @Default(0) double averageSpeedInMetersPerSecond,
+
+    /// The average heart beat during this activity in Beats Per Minute.
+    /// Can be unspecified.
+    double? averageHeartBeatPerMinute,
+
+    /// Additional user notes about this activity.
+    /// Can be empty.
+    @Default('') String notes,
```
Comment (ronw98) [inline lib/domain/entities/activity.dart:45]: rather than settings a default value to an empty string, I think it'd be better to simply make this field nullable

File: lib/domain/entities/activity.dart [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+
+    /// The activity continuous track segments,
+    /// with a pause between each segment.
+    ///
+    /// Can be empty when activity was imported or created manually.
+    @Default([]) List<TrackSegment> trackSegments,
```
Comment (ronw98) [inline lib/domain/entities/activity.dart:51]: actually I think it can be empty when an activity was just started can't it?

File: lib/domain/entities/track_point.dart [hunk 18]
```diff
@@ -0,0 +1,18 @@
...
+part 'track_point.freezed.dart';
+
+/// A point crossed by user at an instant, with associated data.
+@freezed
+class TrackPoint with _$TrackPoint {
+  const TrackPoint._();
+
+  const factory TrackPoint({
+    /// Location with movement data and timestamp.
+    Location? location,
```
Comment (ronw98) [inline lib/domain/entities/track_point.dart:13]: why can this be null?

File: lib/domain/failures.dart [hunk 19]
```diff
@@ -25,4 +25,26 @@ class Failure with _$Failure {
...
+  const factory Failure.database() = _Database;
+
+  /// Failure occurring when the database could not be opened.
+  const factory Failure.databaseNotOpened() = _DatabaseNotOpened;
+
+  /// Failure occurring when the database could not be closed.
+  const factory Failure.databaseNotClosed() = _DatabaseNotClosed;
+
+  /// Failure occurring when something was not successfully saved to database.
+  const factory Failure.databaseNotSaved() = _DatabaseNotSaved;
```
Comment (ronw98) [inline lib/domain/failures.dart:45]: these are `domain` classes and should therefore not be aware of the underlying data source, whether it is a database, a simple file, hardcoded values should not matter.
Failures are more or less related to a use-cases, I think a more appropriate name would be `entitySave`, the given entity could not be saved or `entityDeletion` if the entity was not deleted.

I think `database`,  `databaseNotOpened` and  `databaseNoClosed` should not exist.
Same thing gos for the `adapter` failure actually and that is entirely my fault

File: lib/data/models/activity_model.dart [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+
+@embedded
+class LocationModel {
+  GpsCoordinatesModel gpsCoordinates = GpsCoordinatesModel();
+  double altitudeInMeters = 0;
+  double errorInMeters = 0;
+  double headingInDegrees = 0;
+  double speedInMetersPerSecond = 0;
+  double speedErrorInMetersPerSecond = 0;
+  DateTime? timestamp;
```
Comment (ronw98) [inline lib/data/models/activity_model.dart:58]: I think this should be property of the trakcpoint rather than the location

File: lib/domain/entities/track_point.dart (hunk 18 above)
Comment (ronw98) [reply lib/domain/entities/track_point.dart:13]: also think that the dateTime is more relevant as a property of the TrackPoint rather than the location

File: lib/domain/entities/track_segment.dart [hunk 21]
```diff
@@ -0,0 +1,15 @@
...
+part 'track_segment.freezed.dart';
+
+/// Continuous track segment followed by user, without any pause.
+@freezed
+class TrackSegment with _$TrackSegment {
+  const TrackSegment._();
+
+  const factory TrackSegment({
+    /// All track points followed by user.
+    @Default([]) List<TrackPoint> trackPoints,
```
Comment (ronw98) [inline lib/domain/entities/track_segment.dart:13]: I know we can always find that value using trackpoints but I think a startDate and endDate would make sense here

also, would an automatic pause create multiple segments or would just manual pause create them ?

**Review by ronw98 — CHANGES_REQUESTED** [review]

File: lib/core/modules/isar_module.dart [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+    return newIsarInstance;
+  }
+
+  /// Pre resolved factory that returns the directory for the [Isar] database.
+  @baseIsarDirectory
+  @preResolve
+  Future<Directory> get isarDirectory async {
+    final appDirectory = await getApplicationDocumentsDirectory();
+    return Directory(path.join(appDirectory.path, 'isar'));
+  }
```
Comment (ronw98) [inline lib/core/modules/isar_module.dart:37]: this will be initialized at app startup but I think if the app can't retrieve the documentsDirectory there no point to even try to display anything other than an error screen saying `Well something fucked up so restart the app please`.
So we could just `try-catch` the `configureDependencies` method in `main` and if it fails display said error message.
If it succeeds we'll launch the app normally

File: lib/core/unit_conversion.dart (hunk 4 above)
Comment (BriceCroix) [reply lib/core/unit_conversion.dart:38]: This will go, do not worry, the packages mentionned above handle this better

File: lib/data/adapters/activity_isar_adapter.dart (hunk 2 above)
Comment (BriceCroix) [reply lib/data/adapters/activity_isar_adapter.dart:21]: There are also timezone-aware Datetime classes, but I think the best option in this case is to create a custom `Datetime` class, that could either store :
- `Date`, `TimeOfDay`, `localTimeOffset`
- `DateTime`, `localTimeOffset`
I don't really have an answer there

Comment (ronw98) [reply lib/data/adapters/activity_isar_adapter.dart:21]: the second solution seems fine to me honestly, `localTimeOffset` should be a double as there are 45" and 30" offsets but apart from that I'm good with this soltion

File: lib/core/unit_conversion.dart (hunk 4 above)
Comment (ronw98) [reply lib/core/unit_conversion.dart:38]: right, I forgot

File: lib/data/adapters/activity_isar_adapter.dart (hunk 2 above)
Comment (BriceCroix) [reply lib/data/adapters/activity_isar_adapter.dart:21]: Or we could simply decide to let things as-are, and in the HMI display the local timezone. To keep going with my example, I would see "Activity done at 5am Eastern Standard Time".

I have dealt with timezones before and it is a complete nightmare, the less we do things about it, the better

File: lib/data/models/activity_model.dart (hunk 11 above)
Comment (BriceCroix) [reply lib/data/models/activity_model.dart:46]: I the user decides to track an activity without enabling the GPS, we should be able to store timestamps and heart rates (and other in the future). But the current implementation prevents that since the timestamp is in the `Location` class. I do not like that at all, but I originally did that to copy the implementation of geolocator's position.
`Location` may need to be broken down to `Location` with all position-related fields, and `TimedLocation`, that inherits `Location` or has a `Location` field. This way we could use this modified `Location?` class in `TrackPoint`, and still get `TimedLocation` from the location repository.

Yes that does not answer your question, the original goal was that : to be able not to have GPS data in `TrackPoint`, but as of now doing this prevents from having timestamps.

File: lib/data/models/activity_model.dart (hunk 20 above)
Comment (BriceCroix) [reply lib/data/models/activity_model.dart:58]: Yes I am seeing this now after writing my novel above, but I see that we agree x)

File: lib/domain/entities/activity.dart (hunk 12 above)
Comment (BriceCroix) [reply lib/domain/entities/activity.dart:30]: Yes I see that my comment is unclear, when I say accounting for pauses, I mean that the pauses duration is subtracted from the total time. The `durationIncludingPauses` can be recovered using `stopTime-startTime`

File: lib/domain/entities/activity.dart (hunk 13 above)
Comment (BriceCroix) [reply lib/domain/entities/activity.dart:40]: In the documentation ? I see nos that most fields should be nullable (which to me means unspecified), because if an activity has no location data, no speed or average speed can be set, it does not makes sense to set it to 0, which is what I've done here.

File: lib/domain/entities/activity.dart (hunk 17 above)
Comment (BriceCroix) [reply lib/domain/entities/activity.dart:51]: I don't think so, we should start the activity when we have the initial user location, which allows to create the first `TrackSegment` with a single `TrackPoint`

File: lib/domain/entities/track_point.dart (hunk 18 above)
Comment (BriceCroix) [reply lib/domain/entities/track_point.dart:13]: Yes agree

File: lib/domain/entities/track_segment.dart (hunk 21 above)
Comment (BriceCroix) [reply lib/domain/entities/track_segment.dart:13]: Yes I am not sure a startTime and stopTime field would be relevant here since we can get these values from the Trackpoints, wouldn't getters suffice ?

Comment (BriceCroix) [reply lib/domain/entities/track_segment.dart:13]: An automatic pause would behave the same as a manuel pause there. The aim of the intermediate class between `Activity` and `TrackPoint` is to be able to know where and when pauses happened, to be able to draw different colors on the map.

|======================-----------------------------=====================|
start                                             pause                                   resume                                        stop

File: lib/data/adapters/activity_isar_adapter.dart (hunk 2 above)
Comment (ronw98) [reply lib/data/adapters/activity_isar_adapter.dart]: I say let's just store the `localTimeOffset` and if it happens to be too much of a trouble we'll just ignore it / delete it but I'd rather try to do this best we can

File: lib/data/models/activity_model.dart (hunk 11 above)
Comment (ronw98) [reply lib/data/models/activity_model.dart:51]: Ok so the location can be null depending on the goal of the app
If it's supposed to be an app primarily used to track your location during a physical activity and eventually some other stats to follow along your route I think location should not be null.
If it's supposed to be an app tracking physical activity and location is just a type of data among others then it can be null

In my mind it was designed as the first option, and eventually you could see other data such as heartrate if available.
What I anticipate is that we'll need to perform operations on lists of trackpoints and I feel that enabling `null` location will be troublesome when querying the database, and for a use-case (follow your activity without any speed, or position information) that is not really a need.

Basically I just want to make sure that we are not being to generic and if maybe we could just assume / force some major usages

File: lib/domain/entities/activity.dart (hunk 13 above)
Comment (ronw98) [reply lib/domain/entities/activity.dart]: yes in the documentation.
Unspecified means that setting them manually is optional, it does not necessarilly mean that they are null.
It's really just a wording issue I don't mind that some fields can be null

File: lib/domain/entities/track_segment.dart (hunk 21 above)
Comment (ronw98) [reply lib/domain/entities/track_segment.dart:13]: > Yes I am not sure a startTime and stopTime field would be relevant here since we can get these values from the Trackpoints, wouldn't getters suffice ?

Getters would suffice I guess yes

File: lib/data/adapters/activity_isar_adapter.dart [hunk 23]
```diff
@@ -18,11 +18,13 @@ class ActivityIsarAdapter extends BaseAdapter<Activity, ActivityModel> {
...
       sport: m.sport,
-      // TODO : convert startTime and stopTime back to their original timezone
       startTime: m.startTime,
+      // TODO : handle timezone using startTimeZoneOffsetMinutes and Name
       stopTime: m.stopTime,
       distanceInMeters: m.distanceInMeters,
-      duration: Duration(microseconds: m.durationInMicroSeconds),
+      duration: m.durationInMicroSeconds != null
+          ? Duration(microseconds: m.durationInMicroSeconds!)
+          : null,
```
Comment (ronw98) [inline lib/data/adapters/activity_isar_adapter.dart:27]: this is a personal preference but I think avoiding `!` operators in ternary conditions is more readable

```suggestion
      duration: m.durationInMicroSeconds == null
          ? null
          : Duration(microseconds: m.durationInMicroSeconds!),
```
again, just a personal preference if you prefer having the non-null result as the first part of the ternary condition then fine

File: lib/data/datasources/isar_database_source.dart [hunk 24]
```diff
@@ -1,69 +1,35 @@
...
-      } else {
-        return const Right(null);
-      }
+  /// Save activity to database. Returns the unique ID of the saved object.
+  /// This returned Id should match the one of the given object.
+  Future<Id> saveActivity(ActivityModel model) async {
+    final id = await _isar.writeTxn(() async {
+      return await _isar.activityModels.put(model);
     });
+    return id;
```
Comment (ronw98) [inline lib/data/datasources/isar_database_source.dart]: here, throwing an exception if the result `id` does not match the `model.id` is not a bad idea
It makes sense that the datasource checks the specifics of whether the data has been saved or not and throws if not.
That way the repository does not have to know the specifics of the persistence implementation

File: lib/data/datasources/isar_database_source.dart [hunk 25]
```diff
@@ -1,69 +1,35 @@
...
-      } else {
-        return const Right(null);
-      }
+  /// Returns whether the activity was successfully deleted or not :
+  /// true when deleted.
+  Future<bool> deleteActivity(Id id) async {
+    final deleted = await _isar.writeTxn(() async {
+      return await _isar.activityModels.delete(id);
     });
+    return deleted;
```
Comment (ronw98) [inline lib/data/datasources/isar_database_source.dart]: same thing here, throwing an exception makes sense

File: lib/data/repositories/activity_repository_impl.dart [hunk 26]
```diff
@@ -41,12 +41,8 @@ class ActivityRepositoryImpl
...
   Future<Either<Failure, List<Activity>>> getActivities() async {
     try {
-      final either = await _source.getActivities();
-      return either.fold((failure) {
-        return Left(failure);
-      }, (activities) {
-        return Right(_activityAdapter.modelsToEntities(activities));
-      });
+      final activities = await _source.getActivities();
+      return Right(_activityAdapter.modelsToEntities(activities));
```
Comment (ronw98) [inline lib/data/repositories/activity_repository_impl.dart]: rebase or merge or something on master to change the results using the `Success` format

File: lib/domain/entities/activity.dart [hunk 27]
```diff
@@ -15,34 +15,40 @@ class Activity with _$Activity {
...
+    double? distanceInMeters,
 
-    /// The total duration of this activity, accounting for pauses.
-    @Default(Duration.zero) Duration duration,
+    /// The total duration of this activity, removed pauses durations.
+    /// Null when info is not available.
+    ///
+    /// If an activity starts at 15h, is paused between 15h30 and 15h45 and
+    /// stopped at 16h, the [duration] is 45 minutes.
+    Duration? duration,
```
Comment (ronw98) [inline lib/domain/entities/activity.dart:38]: same thing here, even more for the duration this can only be non null

**Comment by BriceCroix** [comment]
Made important changes :
The class we knew as `Location` is now `TimedLocation`. This is the class returned by the location service. This is not related to the activities.
On the other hand, Activities store `TrackPoint`s, that represents instants of an activity, these use a `Location` field **and** a `Datetime` field, this allows to keep storing Trackpoints with all their data even if the Location service is disabled.

File: lib/core/unit_conversion.dart (hunk 1 above)
Comment (BriceCroix) [reply lib/core/unit_conversion.dart]: this file is gone, units will be handled later on (another PR ?)

File: lib/data/datasources/isar_database_source.dart (hunk 25 above)
Comment (BriceCroix) [reply lib/data/datasources/isar_database_source.dart]: done

File: lib/data/datasources/isar_database_source.dart (hunk 24 above)
Comment (BriceCroix) [reply lib/data/datasources/isar_database_source.dart]: done

File: lib/domain/failures.dart (hunk 19 above)
Comment (BriceCroix) [reply lib/domain/failures.dart]: done

File: lib/data/models/activity_model.dart (hunk 11 above)
Comment (BriceCroix) [reply lib/data/models/activity_model.dart:51]: I prefer to let things as they are, this enables to use the app without any GPS data. The user could for instance track its activity just using its duration. They would only know that that day they practiced this sport for this duration, and I think this is an important feature if their device does not have a GPS or if geolocalisation has been disabled for battery optimization. They could then manually provide the total ran distance if they want to.

**Comment by BriceCroix** [comment]
movna_classes_uml

File: lib/data/models/activity_model.dart [hunk 28]
```diff
@@ -8,11 +8,11 @@ class ActivityModel {
   /// Use startTime as unique identifier.
   Id get id => startTime.microsecondsSinceEpoch;
 
-  DateTime startTime = DateTime.fromMillisecondsSinceEpoch(0);
+  late DateTime startTime;
```
[the remark this answers is not in the document]
Comment (BriceCroix) [reply lib/data/models/activity_model.dart:11]: I honestly don't know either, but I though it was a better idea to actually *require* values rather than putting default ones. If it does not work we'll put default values anyway :)
